src/osp/Active/SysPrefabInit.cpp

```diff
@@ -0,0 +1,201 @@
+
+using osp::restypes::gc_importer;
+
+using osp::phys::EShape;
+
+namespace osp::active
+{
+
+void SysPrefabInit::init_hierarchy(
+        PrefabInitView_t                    prefabInits,
+        Resources const&                    rResources,
+        acomp_storage_t<ACompHierarchy>&    rHier) noexcept
+{
+    for (TmpPrefabInit const& rPrefab : prefabInits)
+    {
+        auto &rImportData = rResources.data_get<osp::ImporterData const>(
+                gc_importer, rPrefab.m_importerRes);
+        auto &rPrefabData = rResources.data_get<osp::Prefabs const>(
+                gc_importer, rPrefab.m_importerRes);
+
+        auto const objects = rPrefabData.m_prefabs[rPrefab.m_prefabId];
+        auto const parents = rPrefabData.m_prefabParents[rPrefab.m_prefabId];
+
+        for (int i = 0; i < objects.size(); ++i)
+        {
+            int const prefabParent = parents[i];
+            ActiveEnt const parent = (prefabParent == -1)
+                    ? rPrefab.m_parent
+                    : rPrefab.m_prefabToEnt[prefabParent];
+
+            ActiveEnt const child = rPrefab.m_prefabToEnt[i];
+
+            SysHierarchy::add_child(rHier, parent, child);
+        }
+    }
+}
+
+void SysPrefabInit::init_transforms(
+        PrefabInitView_t                    prefabInits,
+        Resources const&                    rResources,
+        acomp_storage_t<ACompTransform>&    rTransform) noexcept
+{
+    for (TmpPrefabInit const& rPrefab : prefabInits)
+    {
+        auto const &rImportData = rResources.data_get<osp::ImporterData const>(
+                gc_importer, rPrefab.m_importerRes);
+        auto const &rPrefabData = rResources.data_get<osp::Prefabs const>(
+                gc_importer, rPrefab.m_importerRes);
+
+        auto const objects = rPrefabData.m_prefabs[rPrefab.m_prefabId];
+        auto const parents = rPrefabData.m_prefabParents[rPrefab.m_prefabId];
+
+        for (int i = 0; i < objects.size(); ++i)
+        {
+            Matrix4 const& transform = (parents[i] == -1)
+                    ? *rPrefab.m_pTransform
+                    : rImportData.m_objTransforms[objects[i]];
+            rTransform.emplace(rPrefab.m_prefabToEnt[i], transform);
+        }
+    }
+}
+
+void SysPrefabInit::init_drawing(
+        PrefabInitView_t                    prefabInits,
+        Resources&                          rResources,
+        ACtxDrawing&                        rCtxDraw,
+        ACtxDrawingRes&                     rCtxDrawRes,
+        int                                 materialId) noexcept
+{
+    for (TmpPrefabInit const& rPrefab : prefabInits)
+    {
+        auto &rImportData = rResources.data_get<osp::ImporterData const>(
+                gc_importer, rPrefab.m_importerRes);
+        auto &rPrefabData = rResources.data_get<osp::Prefabs const>(
+                gc_importer, rPrefab.m_importerRes);
+
+        auto const objects = rPrefabData.m_prefabs[rPrefab.m_prefabId];
+
+        for (int i = 0; i < objects.size(); ++i)
```

Comment:
imho, boost::iterator::zip, or (well, with C++23, anyway): https://en.cppreference.com/w/cpp/ranges/zip_view

We could add a dependency on rangesv3, if you want. it would have the same stuff.

Reply:
I like the syntax of zip and ranges, but I'm quite skeptical over how helpful they will really be to this project. There isn't an immediate problem that needs to be solved; I heard the compile time isn't great, and it may make it more difficult to get new members onto the team (the code is already pretty weird but still).